KAFKA-7733: Reset the MockConsumer position according to its reset st…

srdo opened the pull request:
…rategy if the position is behind the beginning offset

The poll method in MockConsumer will now use the configured reset strategy when it encounters a record with an offset below the beginning offset. 

The poll method will also not clear the records list automatically. Letting the consumer position control which records are returned from poll is more similar to how the real consumer works, so e.g. a test can put a message in the consumer, poll for it, seek to the beginning and get the message out again. If tests need to delete the records entirely, they can still do so through a new method.

A few of the methods in MockConsumer were also returning mutable internal maps/lists. The regular consumer returns immutable copies, so I updated the MockConsumer to do the same.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

srdo committed:
KAFKA-7733: Reset the MockConsumer position according to its reset strategy if the position is behind the beginning offset

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -185,24 +186,35 @@ public synchronized void unsubscribe() {
     }
 
+    public synchronized void clearRecords() {
+        this.records.clear();
```
Previously the records were consumed after every poll. Now I think the intent is to treat the records collection as representing the backing log in Kafka. Is that about right? 

Assuming so, I wonder if we can make the representation a little clearer. We currently have separate collections for `beginningOffsets`, `endOffsets`, and `records`. Perhaps we can consolidate all of them. For example, in pseudocode, we could have something like this:

```java
class MockLogData {
  List<ConsumerRecord> log;

  long startOffset() { return log.first.offset(); }
  long endOffset() { return log.last.offset() + 1; }
  List<ConsumerRecord> fetch(long offset) throws OffsetOutOfRangeException; 
}
```

Then we could replace the three collections with a single `Map<TopicPartition, MockLogData>`.

srdo replied:
> Now I think the intent is to treat the records collection as representing the backing log in Kafka. Is that about right?

Yes.

I think something like MockLogData makes sense. We can maybe get rid of `updateBeginningOffsets` and replace it with the ability to delete single offsets from `records`, but we might need to keep `addEndOffset` around. It seems to be intended for testing the case where the code being tested calls `endOffsets` more than once, and the test wants to return something different on each call to simulate records being added concurrently.

I'll try to make something like MockLogData and will let you know when it's ready for another look.

hachikuji reviewed (commented):
Thanks for the patch. I left one comment for consideration.

srdo commented:
@hachikuji I've added something like what you described. Please take a look when you get a chance.

srdo commented:
There are a couple of things that could maybe be improved.
* Maybe the start/end offset for a partition with no messages should be 0 rather than null
* Some tests are using `updateBeginningOffsets/updateEndOffsets` to simulate messages being present, without using `addRecord`. It might be closer to real consumer behavior if we change this and drop the two methods.

What do you think?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -101,12 +98,12 @@ public synchronized void subscribe(Collection<String> topics) {
     public synchronized void subscribe(Pattern pattern, final ConsumerRebalanceListener listener) {
         ensureNotClosed();
-        committed.clear();
+        mockLogs.values().forEach(log -> log.setCommitted(null));
```
To make the mock consumer consistent with the real implementation, it's not clear why we'd want this. Committed offsets persist even after re-subscribing. Do you know if we depend on this logic anywhere?

srdo replied:
I'll try removing it and see what breaks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -181,36 +178,58 @@ public synchronized void unsubscribe() {
+            try {
+                records = getLog(tp).fetch(subscriptions.position(tp));
+            } catch (OffsetOutOfRangeException e) {
+                if (subscriptions.hasDefaultOffsetResetPolicy()) {
```
In the `else` case, should we raise this exception?

srdo replied:
This should already happen, since we call fetch again a few lines below. I can see how it isn't obvious though, so I'll add an `else` that rethrows.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -477,11 +495,8 @@ private void resetOffsetPosition(TopicPartition tp) {
-            return null;
-        }
-        return offsets.size() > 1 ? offsets.remove(0) : offsets.get(0);
+    private MockConsumerLog<K, V> getLog(TopicPartition tp) {
```
If we exposed this, maybe we could get rid of all the forwarding methods (such as `updateBeginningOffsets` and `truncatePartition`?

srdo replied:
Yes. I don't have a strong preference for whether we expose MockConsumerLog, or have forwarding methods. Should I change it to be public and remove the forwarding methods?

hachikuji reviewed (commented):
Thanks @srdo. This is looking pretty good. I left a few suggestions.

hachikuji commented:
> Maybe the start/end offset for a partition with no messages should be 0 rather than null

That seems reasonable.

> Some tests are using updateBeginningOffsets/updateEndOffsets to simulate messages being present, without using addRecord. It might be closer to real consumer behavior if we change this and drop the two methods.

Makes sense.

srdo commented:
@hachikuji I think I've addressed the comments. The changes are in the latest commit, the others were rebased to fix conflicts.

The changes are:
* Removed `updateBeginning/EndOffsets`
* Made the consumer start beginning and end offsets at 0
* Removed the unnecessary `setCommitted(null)` from the `assign` methods on the MockConsumer
* Added a `maxPollRecords` mechanism to the MockConsumer. 

I've updated the tests that were using `updateBeginning/EndOffsets` to fake having messages in the consumer. Most of them were straightforward to replace. The tests in KafkaBasedLogTest want to verify that the code under test call poll until it reaches the end offset. The current code does this by setting the end offset, and not adding the actual messages until a number of polls have happened. The replacement uses maxPollRecords to limit how many messages are returned in each poll, which I think is more similar to real behavior.

KafkaBasedLogTest.testReloadOnStartWithNoNewRecordsPresent wants to start consumption on an empty partition with a non-0 starting and ending offset. The mock log has been updated to support this.

StreamsResetterTest set the end offset, while relying on there being more messages in the consumer after the end offset. I've made it simply add more messages after the endOffset check, which causes the same behavior.

The failing test on Jenkins for Java 11 passes for me locally, it looks like it's just a timeout. The failing test on Java 8 doesn't look like one I've touched.

srdo commented:
@hachikuji Friendly reminder, I think this is ready for review when you get a chance.

srdo commented:
Rebased, squashed and fixed conflicts.

srdo commented:
This is way out of date, and there didn't seem to be much interest. Closing.
